### PR TITLE
feature[next]: Temporaries

### PR DESCRIPTION
**Pull request by tehrengruber**
This PR adds support for automatic creation of temporaries. Right now unstructured domain sizes are fixed at code generation time, but this could easily be lifted. Cartesian domains are symbolic.

Todos (will be addressed in future PRs):
- Scan are not supported properly.
- Many tests on Iterator level are parameterized not only on the backend, but also `LiftMode`. This PR however makes the `LiftMode` a property of the backend (as it should be), which makes the test structure awkward. Due to time constraints and the unclear interface for configuration options of the backend (still in the working in the cleanup project) this will be addressed at a later time.
- Temporaries for expressions of type tuple are at least sup-optimal at this time (some extractions are just unnecessary).
- A proper heuristics on whether it makes sense performance-wise to extract a temporary.

Example transformation:
```
__field_operator_composed_shift_unstructured_flat(__sym_1, out, ____sym_1_size_0, __out_size_0) {
  out ← (
      λ(inp) → ·(↑(λ(it) → ·⟪C2Eₒ, 0ₒ⟫(it)))((↑(λ(it) → ·⟪E2Vₒ, 0ₒ⟫(it)))(inp))
    )(__sym_1) @ u⟨ Cell: [0, __out_size_0) ⟩;
}
```
is transformed into:
```
__field_operator_composed_shift_unstructured_flat(__sym_1, out, ____sym_1_size_0, __out_size_0) {
  _tmp_1 = temporary(domain=u⟨ Edge: [0, 18) ⟩, dtype=int32);
  __field_operator_composed_shift_unstructured_flat(
    __sym_1,
    out,
    ____sym_1_size_0,
    __out_size_0,
    _tmp_1
  ) {
    _tmp_1 ← (λ(it) → ·⟪E2Vₒ, 0ₒ⟫(it))(__sym_1) @ u⟨ Edge: [0, 18) ⟩;
    out ← (λ(_tmp_1) → ·⟪C2Eₒ, 0ₒ⟫(_tmp_1))(_tmp_1) @ u⟨ Cell: [0, __out_size_0) ⟩;
  };
  __field_operator_composed_shift_unstructured_flat(
    __sym_1, out, ____sym_1_size_0, __out_size_0, _tmp_1
  );
}
```

### CODE REVIEW

File: src/gt4py/next/iterator/tracing.py [hunk 1]
```diff
@@ -172,6 +174,7 @@ def trace_function_call(fun, *, args=None):
     body = fun(*list(args))
 
     if isinstance(body, tuple):
+        # TODO(tehrengruber): This fails for nested tuples.
```
Comment (havogt) [inline src/gt4py/next/iterator/tracing.py:177]: Can you add the why? Not obvious.

File: src/gt4py/next/iterator/tracing.py [hunk 2]
```diff
@@ -172,6 +174,7 @@ def trace_function_call(fun, *, args=None):
...
         return _f("make_tuple", *tuple(make_node(b) for b in body))
     else:
         return make_node(body) if body is not None else None
```
Comment (havogt) [inline src/gt4py/next/iterator/tracing.py:180]: Actually why does this not work?

```suggestion
    return make_node(body) if body is not None else None
```

File: src/gt4py/next/iterator/tracing.py [hunk 3]
```diff
@@ -19,6 +19,7 @@
 from gt4py.eve import Node
```
Comment (havogt) [inline src/gt4py/next/iterator/tracing.py:19]: Maybe can we extract yet another pr e.g. everything in iterator except in transforms. I would like to be able to focus on the core of the PR without all the noise around. Let's discuss it.

File: src/gt4py/next/iterator/transforms/collapse_tuple.py [hunk 4]
```diff
@@ -15,16 +15,15 @@
 from dataclasses import dataclass
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/collapse_tuple.py:15]: note to self: not reviewed, probably all changes in other PR or already merged?

File: src/gt4py/next/iterator/transforms/cse.py [hunk 5]
```diff
@@ -25,6 +25,8 @@
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/cse.py:25]: note to self: probably all changes are in the other PR

File: src/gt4py/next/iterator/transforms/inline_lambdas.py [hunk 6]
```diff
@@ -110,12 +125,22 @@ def new_name(name):
 class InlineLambdas(NodeTranslator):
     """Inline lambda calls by substituting every argument by its value."""
 
+    PRESERVED_ANNEX_ATTRS = ("type",)
+
     opcount_preserving: bool
 
     force_inline_lift: bool
 
+    force_inline_trivial_lift: bool
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/inline_lambdas.py:134]: let's describe here why we wouldn't want to do that always.

Comment (havogt) [reply src/gt4py/next/iterator/transforms/inline_lambdas.py:134]: and why is it here and not in lift inliner?

File: src/gt4py/next/iterator/transforms/inline_lifts.py [hunk 7]
```diff
@@ -108,15 +109,25 @@ def _transform_and_extract_lift_args(
...
-        if predicate is None:
-            self.predicate = lambda _1, _2: True
-        else:
-            self.predicate = predicate
+    class Flag(enum.IntEnum):
+        PROPAGATE_SHIFT = 1
+        INLINE_TRIVIAL_DEREF_LIFT = 2
+        INLINE_DEREF_LIFT = 2 + 4
+        PROPAGATE_CAN_DEREF = 8
+        INLINE_LIFTED_ARGS = 16
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/inline_lifts.py:121]: Same question as above? Let's document why we need the different ways.

File: src/gt4py/next/iterator/transforms/symbol_ref_utils.py [hunk 8]
```diff
@@ -13,44 +13,59 @@
 # SPDX-License-Identifier: GPL-3.0-or-later
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/symbol_ref_utils.py:13]: Again for my sanity, maybe we can extract this change into a separate PR. Do we have tests for these utils?

File: src/gt4py/next/iterator/transforms/trace_shifts.py [hunk 9]
```diff
@@ -11,23 +11,27 @@
 # distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/trace_shifts.py:11]: not reviewed yet. separate pr? 🙃

**Review by havogt — CHANGES_REQUESTED** [review]
Let's discuss how much work it is to split it into several PRs, sucht that I can review the temporaries part in insolation.

File: src/gt4py/next/iterator/tracing.py (hunk 1 above)
Comment (tehrengruber) [reply src/gt4py/next/iterator/tracing.py:177]: Fixed in #1317

File: src/gt4py/next/iterator/tracing.py (hunk 2 above)
Comment (tehrengruber) [reply src/gt4py/next/iterator/tracing.py:180]: Fixed in #1317

File: src/gt4py/next/iterator/transforms/global_tmps.py [hunk 10]
```diff
@@ -11,18 +11,23 @@
 # distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/global_tmps.py:11]: please remove this file from the mypy exclusion list. enjoy! ;)

File: src/gt4py/next/iterator/transforms/inline_lambdas.py [hunk 11]
```diff
@@ -110,12 +125,22 @@ def new_name(name):
 class InlineLambdas(NodeTranslator):
     """Inline lambda calls by substituting every argument by its value."""
 
+    PRESERVED_ANNEX_ATTRS = ("type",)
+
     opcount_preserving: bool
 
     force_inline_lift: bool
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/inline_lambdas.py:132]:
```suggestion
    force_inline_lift_args: bool
```

File: src/gt4py/next/iterator/transforms/pass_manager.py [hunk 12]
```diff
@@ -93,12 +93,26 @@ def apply_common_transforms(
...
     else:
-        ir = InlineLambdas.apply(
-            ir, opcount_preserving=True, force_inline_lift=(lift_mode == LiftMode.FORCE_INLINE)
-        )
+        for _ in range(10):
+            inlined = InlineLambdas.apply(
+                ir,
+                opcount_preserving=True,
+                force_inline_lift=(lift_mode == LiftMode.FORCE_INLINE),
+                force_inline_trivial_lift=True,
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/pass_manager.py:101]: Document why `force_inline_trivial_lift`

File: src/gt4py/next/iterator/transforms/global_tmps.py [hunk 13]
```diff
@@ -113,9 +119,53 @@ def pformat_FencilWithTemporaries(
...
+
+
+def temporary_extraction_predicate(expr: ir.Node, num_occurences: int) -> bool:
+    """Determine if `expr` is an applied lift that should be extracted as a temporary."""
+    if not (
+        isinstance(expr, ir.FunCall)
+        and isinstance(expr.fun, ir.FunCall)
+        and expr.fun.fun == ir.SymRef(id="lift")
+    ):
+        return False
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/global_tmps.py:155]: Let's do the refactoring of extracting `is_applied_lift` to a shared location from all passes. (if you like to do it in a follow up pr, fine, but then add a TODO

File: src/gt4py/next/iterator/transforms/global_tmps.py [hunk 14]
```diff
@@ -127,56 +177,122 @@ def split_closures(node: ir.FencilDefinition) -> FencilWithTemporaries:
...
-        popped_stencil = PopupTmps().visit(wrapped_stencil)
+        closure_stack: list[ir.StencilClosure] = [closure]
+        while closure_stack:
+            current_closure: ir.StencilClosure = closure_stack.pop()
+
+            if current_closure.stencil == im.ref("deref"):
+                closures.append(current_closure)
+                continue
+
+            # TODO(tehrengruber): extracting from a scan is untested.
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/global_tmps.py:196]: Maybe should we disable the scan case for now until we have a test?

File: src/gt4py/next/iterator/transforms/global_tmps.py [hunk 15]
```diff
@@ -127,56 +177,122 @@ def split_closures(node: ir.FencilDefinition) -> FencilWithTemporaries:
...
-            domain = AUTO_DOMAIN
+
+            if extracted_lifts:
+                for tmp_sym, lift_expr in extracted_lifts.items():
+                    # make sure the applied lift is not capturing anything except of closure params
+                    used_symbol_refs = collect_symbol_refs(lift_expr)
+                    assert not (
+                        set(used_symbol_refs)
+                        - {param.id for param in current_closure_stencil.params}
+                    )
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/global_tmps.py:219]: nit: I'd prefer something like `assert _applied_lift_captures_only_closure_params(lift_expr, current_closure_stencil.params)` because I don't like comments that could be code.

File: src/gt4py/next/iterator/transforms/global_tmps.py [hunk 16]
```diff
@@ -127,56 +177,122 @@ def split_closures(node: ir.FencilDefinition) -> FencilWithTemporaries:
...
+
+                    assert isinstance(lift_expr, ir.FunCall) and isinstance(
+                        lift_expr.fun, ir.FunCall
+                    )
+
+                    # make sure the arguments to the applied lift are only symbols
+                    # (otherwise we would need to canonicalize using `canonicalize_applied_lift`
+                    # this doesn't seem to be necessary right now as we extract the lifts
+                    # in post-order of the tree)
+                    assert all(isinstance(arg, ir.SymRef) for arg in lift_expr.args)
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/global_tmps.py:229]: does it mean the `canonicalize_applied_lift` is currently not used? actually here is the comment (first line) that I want in the `canonicalize_applied_lift`.

File: src/gt4py/next/iterator/transforms/global_tmps.py [hunk 17]
```diff
@@ -127,56 +177,122 @@ def split_closures(node: ir.FencilDefinition) -> FencilWithTemporaries:
...
+                    stencil_arg_map = {
+                        param.id: arg
+                        for param, arg in zip(
+                            current_closure_stencil.params
+                            if not is_scan
+                            else current_closure_stencil.params[1:],
+                            current_closure.inputs,
+                            strict=True,
+                        )
+                    }
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/global_tmps.py:243]: extract function (goal is to make it possible to get an overview of this huge function/loop, by removing details)

File: src/gt4py/next/iterator/transforms/global_tmps.py [hunk 18]
```diff
@@ -373,103 +348,179 @@ def _max_domain_sizes_by_location_type(offset_provider: Mapping[str, Any]) -> di
...
-                assert isinstance(arg, ir.FunCall) and arg.fun == ir.SymRef(id="named_range")
-                axis = arg.args[0].value
-                ranges.setdefault(axis, []).append(arg)
-    return ranges
+@dataclasses.dataclass
+class SymbolicRange:
+    start: ir.Expr
+    stop: ir.Expr
 
+    def translate(self, distance):
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/global_tmps.py:356]:
```suggestion
    def translate(self, distance: ir.Expr) -> SymbolicRange:
```

File: src/gt4py/next/iterator/transforms/global_tmps.py [hunk 19]
```diff
@@ -373,103 +348,179 @@ def _max_domain_sizes_by_location_type(offset_provider: Mapping[str, Any]) -> di
...
-            domains[param] = _unstructured_domain(
-                loctype, horizontal_size, vertical_ranges.values()
+        for param, shift_chains in local_shifts.items():
+            assert isinstance(param, str)
+            consumed_domains: list[SymbolicDomain] = (
+                [SymbolicDomain.from_expr(domains[param])] if param in domains else []
             )
+            for shift_chain in shift_chains:
+                consumed_domain = SymbolicDomain.from_expr(domain)
+                for shift in zip(shift_chain[::2], shift_chain[1::2], strict=True):
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/global_tmps.py:431]: this is hard to read if you don't know the internals very well -> `group_offsets(offsets: Sequence[Offset]) -> tuple[tag,int]`

File: src/gt4py/next/iterator/transforms/global_tmps.py [hunk 20]
```diff
@@ -373,103 +348,179 @@ def _max_domain_sizes_by_location_type(offset_provider: Mapping[str, Any]) -> di
...
+                    start = functools.reduce(
+                        lambda current_expr, el_expr: im.call("minimum")(current_expr, el_expr),
+                        [consumed_domain.ranges[dim].start for consumed_domain in consumed_domains],
+                    )
+                    stop = functools.reduce(
+                        lambda current_expr, el_expr: im.call("maximum")(current_expr, el_expr),
+                        [consumed_domain.ranges[dim].stop for consumed_domain in consumed_domains],
+                    )
+                    param_domain_ranges[dim] = SymbolicRange(start, stop)
+                assert domain.fun in [im.ref("unstructured_domain"), im.ref("cartesian_domain")]
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/global_tmps.py:469]: this looks also like a nice function to extract.

File: src/gt4py/next/iterator/transforms/inline_lambdas.py [hunk 21]
```diff
@@ -21,10 +21,23 @@
...
 
-def inline_lambda(
+def _is_applied_lift(arg: ir.Node) -> bool:
+    """Match expressions of the form `lift(λ(...) → ...)(...)`."""
+    return (
+        isinstance(arg, ir.FunCall)
+        and isinstance(arg.fun, ir.FunCall)
+        and isinstance(arg.fun.fun, ir.SymRef)
+        and arg.fun.fun.id == "lift"
+    )
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/inline_lambdas.py:31]: ah here it is ;)

File: tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py [hunk 22]
```diff
@@ -895,6 +918,15 @@ def program_domain_tuple(
...
 def test_where_k_offset(cartesian_case):
+    if cartesian_case.backend in [
+        gtfn_cpu.run_gtfn,
+        gtfn_cpu.run_gtfn_imperative,
+        gtfn_cpu.run_gtfn_with_temporaries,
+    ]:
+        pytest.xfail("IndexFields are not supported yet.")
+    if cartesian_case.backend == dace_iterator.run_dace_iterator:
+        pytest.xfail("Not supported in DaCe backend: index fields")
+
```
Comment (havogt) [inline tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py:929]: merge conflict?

File: tests/next_tests/integration_tests/feature_tests/iterator_tests/test_horizontal_indirection.py [hunk 23]
```diff
@@ -101,6 +102,9 @@ def test_direct_offset_for_indirection(program_processor):
     if program_processor == run_dace_iterator:
         pytest.xfail("Not supported in DaCe backend: shift offsets not literals")
 
+    if program_processor == gtfn_cpu.run_gtfn_with_temporaries:
+        pytest.xfail("Dynamic offsets not supported in temporaries pass.")
```
Comment (havogt) [inline tests/next_tests/integration_tests/feature_tests/iterator_tests/test_horizontal_indirection.py:106]: what's the strategy to resolve that? I guess the pass should just not try to extract, but should work, right?

File: tests/next_tests/integration_tests/multi_feature_tests/ffront_tests/test_icon_like_scan.py [hunk 24]
```diff
@@ -230,6 +234,8 @@ def test_solve_nonhydro_stencil_52_like_z_q(test_setup, fieldview_backend):
 
 
 def test_solve_nonhydro_stencil_52_like_z_q_tup(test_setup, fieldview_backend):
+    if fieldview_backend in [gtfn_cpu.run_gtfn_with_temporaries]:
+        pytest.xfail("Needs implementation of scan projector.")
```
Comment (havogt) [inline tests/next_tests/integration_tests/multi_feature_tests/ffront_tests/test_icon_like_scan.py:238]: merge conflict or is the project the reason why it doesn't work?

File: tests/next_tests/integration_tests/multi_feature_tests/iterator_tests/test_hdiff.py [hunk 25]
```diff
@@ -78,7 +78,12 @@ def hdiff(inp, coeff, out, x, y):
 
 def test_hdiff(hdiff_reference, program_processor_no_dace_exec, lift_mode):
     program_processor, validate = program_processor_no_dace_exec
-    if program_processor == run_gtfn or program_processor == run_gtfn_imperative:
+    if program_processor in [
+        gtfn_cpu.run_gtfn,
+        gtfn_cpu.run_gtfn_imperative,
+        gtfn_cpu.run_gtfn_with_temporaries,
+    ]:
+        # TODO(tehrengruber): check if still true
```
Comment (havogt) [inline tests/next_tests/integration_tests/multi_feature_tests/iterator_tests/test_hdiff.py:86]: This TODO should be resolved in this PR, either it's true, then add a more precise description, or it works and remove the xfail.

File: tests/next_tests/integration_tests/multi_feature_tests/iterator_tests/test_vertical_advection.py [hunk 26]
```diff
@@ -119,15 +119,25 @@ def fen_solve_tridiag2(i_size, j_size, k_size, a, b, c, d, x):
...
+        program_processor
+        in [
+            gtfn_cpu.run_gtfn,
+            gtfn_cpu.run_gtfn_imperative,
+            gtfn_cpu.run_gtfn_with_temporaries,
+            gtfn_format_sourcecode,
+        ]
+        and lift_mode == LiftMode.FORCE_INLINE
+    ):
         pytest.xfail("gtfn does only support lifted scans when using temporaries")
```
Comment (havogt) [inline tests/next_tests/integration_tests/multi_feature_tests/iterator_tests/test_vertical_advection.py:135]:
```suggestion
        pytest.skip("gtfn does only support lifted scans when using temporaries")
```

File: tests/next_tests/unit_tests/iterator_tests/test_type_inference.py [hunk 27]
```diff
@@ -18,6 +18,31 @@
...
 
 
+def test_unsatisfiable_constraints():
+    a = ir.Sym(id="a", dtype=("float32", False))
+    b = ir.Sym(id="b", dtype=("int32", False))
+
+    testee = im.lambda_(a, b)(im.plus("a", "b"))
+
+    # TODO(tehrengruber): For whatever reason the ordering in the error message is not
+    #  deterministic. Ignoring for now, as we want to refactor the type inference anyway.
```
Comment (havogt) [inline tests/next_tests/unit_tests/iterator_tests/test_type_inference.py:28]: Maybe update the comment to reflect why we think the ordering is not deterministic.

File: tests/next_tests/unit_tests/iterator_tests/test_type_inference.py [hunk 28]
```diff
@@ -724,6 +749,9 @@ def test_stencil_closure():
             dtype=ti.TypeVar(idx=0),
             size=ti.Column(),
             current_loc=ti.ANYWHERE,
+            # TODO: remove comment below
```
Comment (havogt) [inline tests/next_tests/unit_tests/iterator_tests/test_type_inference.py:752]: resolve this TODO ;)

**Review by havogt — CHANGES_REQUESTED** [review]

File: src/gt4py/next/iterator/transforms/global_tmps.py (hunk 14 above)
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/global_tmps.py:196]: I've added a test.

File: src/gt4py/next/iterator/transforms/global_tmps.py (hunk 16 above)
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/global_tmps.py:229]: Yes. It looked too nice and untrivial to throw it away as I am not a 100% certain it is not needed. I can remove it if you like.

File: src/gt4py/next/iterator/transforms/global_tmps.py (hunk 15 above)
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/global_tmps.py:219]: I don't see that point here to be honest. The comment can rod the same way the name of the function can. If I change the code without updating the comment that appears as likely as changing a function without updating its name (otherwise one is discriminating comments :-P). Your choice, but the very long function name in comparison to its content seems a little odd.

File: tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py (hunk 22 above)
Comment (tehrengruber) [reply tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py:929]: Yes

File: tests/next_tests/integration_tests/feature_tests/iterator_tests/test_horizontal_indirection.py (hunk 23 above)
Comment (tehrengruber) [reply tests/next_tests/integration_tests/feature_tests/iterator_tests/test_horizontal_indirection.py:106]: As far as I remember we only need to extend the `TraceShifts` pass. Shouldn't be very time intensive.

File: src/gt4py/next/iterator/transforms/global_tmps.py (hunk 14 above)
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/global_tmps.py:217]: TODO: remove comment

File: src/gt4py/next/iterator/transforms/global_tmps.py (hunk 15 above)
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/global_tmps.py:240]: Conclusion: make function :-P

File: tests/next_tests/integration_tests/feature_tests/iterator_tests/test_horizontal_indirection.py (hunk 23 above)
Comment (tehrengruber) [reply tests/next_tests/integration_tests/feature_tests/iterator_tests/test_horizontal_indirection.py:106]: Idea: TraceShift returns something like `(SomeDimension, DynamicShiftDistance)`. Temporary pass itself should just work.

File: tests/next_tests/unit_tests/iterator_tests/test_type_inference.py (hunk 28 above)
Comment (havogt) [reply tests/next_tests/unit_tests/iterator_tests/test_type_inference.py]: why is it still here?

File: tests/next_tests/integration_tests/feature_tests/iterator_tests/test_horizontal_indirection.py (hunk 23 above)
Comment (havogt) [reply tests/next_tests/integration_tests/feature_tests/iterator_tests/test_horizontal_indirection.py:106]: Can you add this comment to the code.

File: tests/next_tests/unit_tests/iterator_tests/test_type_inference.py (hunk 28 above)
Comment (havogt) [reply tests/next_tests/unit_tests/iterator_tests/test_type_inference.py]: sorry, was looking at the outdated code because I clicked somewhere...

File: tests/next_tests/integration_tests/feature_tests/iterator_tests/test_horizontal_indirection.py (hunk 23 above)
Comment (tehrengruber) [reply tests/next_tests/integration_tests/feature_tests/iterator_tests/test_horizontal_indirection.py:106]: Thinking about this again I was wrong in assuming this would be easy to fix. I've added my reasoning as a comment in `TraceShift`.

**Review by havogt — APPROVED** [review]
